src/python/pants/backend/experimental/python/lockfile_metadata.py

```diff
@@ -0,0 +1,75 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import hashlib
+from dataclasses import dataclass
+
+from pants.backend.python.util_rules.interpreter_constraints import InterpreterConstraints
+from pants.util.ordered_set import FrozenOrderedSet
+
+BEGIN_LOCKFILE_HEADER = b"# --- BEGIN PANTS LOCKFILE METADATA: DO NOT EDIT OR REMOVE ---"
+END_LOCKFILE_HEADER = b"# --- END PANTS LOCKFILE METADATA ---"
+
+
+@dataclass
+class LockfileMetadata:
+    invalidation_digest: str | None
+
+
+def invalidation_digest(
+    requirements: FrozenOrderedSet[str], interpreter_constraints: InterpreterConstraints
+) -> str:
+    """Returns an invalidation digest for the given requirements and interpreter constraints."""
+    m = hashlib.sha256()
+    for requirement in requirements:
+        m.update(requirement.encode("utf-8"))
+    for constraint in interpreter_constraints:
```

Comment:
CPython is both a valid interpreter constraint and a valid requirement (https://pypi.org/project/cPython/). As such the lone requirement `cpython` will hash the same as the lone IC cpython. Its probably the case that we short circuit in higher layers for 0 requirements, but that's not checked here / it's non-local behavior we're relying on. Perhaps pad the hash with `m.update(b'@ICs@')` - which is an invalid requirement string - before this for loop if perf has been a measured concern, but really, just storing all this in a `{"reqs; [...], "ics": []}` and hashing the json encoding of that dict is probably the most robust, readable and performant enough.

Reply:
Thank you, this is a good catch. I'll add a test case and update to use the JSON repr of the two.

Reply:
@jsirois Note that this is a WIP PR for the purposes of discussion of in-development stuff with @Eric-Arellano. The original implementation of this function is over in #12427 and I'll be resolving the problem over there.